File: 21553 - Sai Nandan D/Week 2/1541_Minimum Insertions to Balance a Parentheses String.py
class Solution:
    def minInsertions(self, s: str) -> int:
        pairs, ones = 0, 0
        i, n = 0, len(s)
        
        while i < n:
            if s[i] == '(': pairs += 1                
            else:
                if i == n-1 or s[i+1] != ')': ones += 1
                else: i += 1                
                if not pairs: ones += 1
                else: pairs -= 1                
            i += 1
            
        return 2 * pairs + ones

File: 21551-naveen-siva-kumar/week2/1541-minimum-insertions-to-balance-a-parentheses-string/1541-minimum-insertions-to-balance-a-parentheses-string.cpp

class Solution {
public:
    int minInsertions(string s) {
        int n=s.size();
        int cost=0, open=0;
        
        for(int i=0;i<n;i++){
            
            if(s[i] == '(')
                open++;
            else{
               
                if(open==0){
                    cost++;                    
                }
                else{
					
                    open--;
                }
               
                if(i==n-1 or s[i+1] != ')')cost++;
                else i++;
            }
        }
        
        cost += 2*(open); //this is for all those ( paranthesis still not closed
        return cost;
    }
};

